src/compiler/propagation/type_propagator.cc

```diff
@@ -1106,15 +1106,27 @@ static TypeScope* process(TypeScope* scope, uint8* bcp, std::vector<Worklist*>&
   OPCODE_END();
 
   OPCODE_BEGIN(NON_LOCAL_RETURN);
+    int level = stack->local(0).block()->level();
     stack->pop();  // Pop block.
-    method->ret(propagator, stack);
+    if (level == 0) {
+      method->ret(propagator, stack);
+    } else {
+      BlockTemplate* block = worklists[level]->block();
```

Comment:
Add a comment, reminding readers that the block-level and the work-list level are synchronized.

Reply:
Done.